### PR TITLE
XWIKI-15653 Prevent users when refactoring are done on pages with XClass

### PR DESCRIPTION
**Pull request by surli**
* Prevent users when they try renaming a page

### CODE REVIEW

**Comment by surli** [comment]
This is a first step for fixing XWIKI-15653: trying to apply it on renaming first. 

@atallahade @mflorea @tmortagne I'd like a review before continuing for the "copy/delete" refactoring. The idea is to put the main macro in refactoring_macros.vm and to apply the same kind of process for the force.

File: xwiki-platform-core/xwiki-platform-web/src/main/webapp/templates/rename.vm [hunk 1]
```diff
@@ -84,4 +114,44 @@
   </div>## mainContentArea
   </div>## main
   #template("endpage.vm")
+
+## There is class being used and renaming not forced
+#else
+  #template("startpage.vm")
+  <div class="main">
+    <div id="mainContentArea">
+    ##
```
Comment (atallahade) [inline xwiki-platform-core/xwiki-platform-web/src/main/webapp/templates/rename.vm]: Is this line required?

File: xwiki-platform-core/xwiki-platform-web/src/main/webapp/templates/rename.vm [hunk 2]
```diff
@@ -84,4 +114,44 @@
...
+      $response.setStatus(423)
```
Comment (atallahade) [inline xwiki-platform-core/xwiki-platform-web/src/main/webapp/templates/rename.vm]: You could put a comment (like you did below) to specify what this error means.

File: xwiki-platform-core/xwiki-platform-extension/xwiki-platform-extension-handlers/xwiki-platform-extension-handler-xar/src/main/resources/ApplicationResources.properties [hunk 3]
```diff
@@ -113,3 +113,6 @@ extension.xar.editwarning.list=The current page belong to the following extensio
 extension.xar.editwarning.explanation=If you modify this page, the extensions might not work anymore.
 extension.xar.editdeny.list=The current page belong to the following extensions:
 extension.xar.editdeny.explanation=You are not allowed to modify protected extension pages.
+
+extension.xar.renamewarning=The following pages contains XClass that are being used:
```
[the remark this answers is not in the document]
Comment (mflorea) [reply xwiki-platform-core/xwiki-platform-extension/xwiki-platform-extension-handlers/xwiki-platform-extension-handler-xar/src/main/resources/ApplicationResources.properties]: Yes, the best practice is to use camel case.

File: xwiki-platform-core/xwiki-platform-extension/xwiki-platform-extension-handlers/xwiki-platform-extension-handler-xar/src/main/resources/ApplicationResources.properties [hunk 4]
```diff
@@ -113,3 +113,6 @@ extension.xar.editwarning.list=The current page belong to the following extensio
...
+extension.xar.renamedeny=This page or one of its children contains XClass that are being used. Ask an advanced user to force the renaming.
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-extension/xwiki-platform-extension-handlers/xwiki-platform-extension-handler-xar/src/main/resources/ApplicationResources.properties]: This message could be improved also (see above).

**Comment by surli** [comment]
> I'm not sure if this is the right approach. @tmortagne has put the code that prevents the deletion of extension pages inside the refactoring job.

that's why I was asking to review before going further :)
I actually reuse the approach existing for forcing the edit of pages: so what I did here is completely based on what already exists in `edit.vm`.
Now I saw the code which is used to prevent deletion of extension pages, and AFAIR it's based on job questions, so it happens after the job already began: not sure how it can be used to forbid simple user to do the refactoring.

IMHO they should see the warning before going to the renaming form.

File: xwiki-platform-core/xwiki-platform-web/src/main/webapp/templates/rename.vm [hunk 5]
```diff
@@ -1,19 +1,49 @@
...
+  #set ($query = $query.setWiki($spaceReference.wikiReference.name))
+  #set ($localSpaceReference = $services.model.serialize($spaceReference, 'local'))
+  #set ($query = $query.bindValue('space', $localSpaceReference))
+  #set ($spacePrefix = $localSpaceReference.replaceAll('([%_/])', '/$1'))
+  #set ($query = $query.bindValue('spacePrefix', "${spacePrefix}.%"))
+  #set ($return = $NULL)
+  #setVariable("$return" $query.execute())
+#end
+
+#set ($spaceReference = $doc.documentReference.parent)
```
[the remark this answers is not in the document]
Comment (surli) [reply xwiki-platform-core/xwiki-platform-web/src/main/webapp/templates/rename.vm]: Agreed for the first one.
Now for the second one: it's on purpose, I want to display the warning/error if the page contains nested page with XClass. It was part of my proposal.

**Comment by tmortagne** [comment]
> @tmortagne has put the code that prevents the deletion of extension pages inside the refactoring job.

s/@tmortagne/@gdelhumeau/ actually, I just added `rename` support and XAR entry types :)

File: xwiki-platform-core/xwiki-platform-web/src/main/webapp/templates/rename.vm (hunk 5 above)
[the remark this answers is not in the document]
Comment (mflorea) [reply xwiki-platform-core/xwiki-platform-web/src/main/webapp/templates/rename.vm]: >  if the page contains nested page with XClass

Yes, but the user can choose to not rename the nested pages. There's an option to "Preserve children" on the rename form that can be unchecked. If you do this then only the current page is renamed so you should check only if the current page is a class being used (same as with the case of a terminal page).

**Comment by surli** [comment]
So just discussed that about @vmassol and finally we agreed it makes more sense to use the same behaviour as in deleting pages that belongs to an extension.
The question is: how this mechanism currently works? Does it do the job until a question should be asked, or are the questions asked before the job start?

Because here we don't want a refactoring to start at all if a user has checked a page that contains a xclass.

**Comment by tmortagne** [comment]
By the way the rationale to put it in the Job was that analyzing all that might take a while and handling slow stuff is supposed to be what jobs are for.

>  The question is: how this mechanism currently works?

`org.xwiki.refactoring.internal.job.AbstractEntityJobWithChecks`

1. The job gathers all the "concerned entities"
2. It send a DocumentsDeletingEvent with the job instance and a Map<EntityReference, EntitySelection>. The listeners are free to modify each EntitySelection, also DocumentsDeletingEvent is cancellable
3. The XAR extension handler have a listener which receive this event, modify the default set of each EntitySelection depending on what is dangerous or not from extension point of view and inject a question in the current job to give a chance to the user to decide. (`org.xwiki.extension.xar.internal.delete.DocumentsDeletingListener`)

So yes the Job is started technically but nothing is actually done before XAR handler's question (if any) is answered for example.

For your use cases there is several choices:
* add class stuff to the XAR extension handler question but it's not very clean since it's not really a XAR extension thing
* implement some listener (in xwiki-platform-refactoring-default I guess or some other module if we think it's not standard enough) which does the same than the XAR handler but for classes (and potentially other standard stuff later)

**Comment by surli** [comment]
Thanks @tmortagne for the explanation here. I was crawling the code and indeed the second option looks better. 
Closing this PR now as I'm going in a completely different direction.

**Pull request closed without merging**